File: PROGRAMMING/MATLAB CODE/leadsensor.m
clc;
clear;
close all;
%1.Simulate Lead sensor data
fs=10;
t=0:1/fs:100;   
true_lead=15+3*sin(0.1*t)+0.5*randn(size(t));
%Add random sensor noise(simulating measurement error)
noise=1*randn(size(t));
lead_noisy=true_lead+noise;
%2.Moving Average Filter
window_size=5; %Number of samples in windows
b=(1/window_size)*ones(1,window_size);
a=1;
lead_moving_avg=filter(b,a,lead_noisy);
%3.Median Filter
windowSize_med=5;
lead_median=medfilt1(lead_noisy,windowSize_med);
%4.Butterworth Low-pass Filter      
cutoff_freq=0.1; %Normalized cutoff frequency
[b_butter,a_butter]=butter(4,cutoff_freq);
lead_butter=filter(b_butter,a_butter,lead_noisy);
%5.Plot results
figure('Name','Lead Sensor Data Filtering','NumberTitle','off');
subplot(4,1,1);
plot(t,lead_noisy,'r');
title('Unfiltered Lead Sensor Data(Noisy)');
xlabel('Time (s)');
ylabel('Lead Concentration (ppb)');
grid on;
subplot(4,1,2);
plot(t,lead_moving_avg,'b');
title('Lead Sensor Data - Moving Average Filter');
xlabel('Time (s)');
ylabel('Lead Concentration (ppb)');
grid on;
subplot(4,1,3);
plot(t,lead_median,'g');
title('Filtered with Median filter');
xlabel('Time (s)');
ylabel('Lead Concentration (ppb)');
grid on;
subplot(4,1,4);
plot(t,lead_butter,'m');
title('Filtered with Butterworth Low-Pass Filter (filtfilt)');
xlabel('Time (s)');
ylabel('Lead Concentration (ppb)'); 
grid on;
sgtitle('Water Treatment:Lead Sensor Filtering Comparison');
xlabel('Time (s)');
ylabel('Lead Concentration (ppb)');
grid on;

File: PROGRAMMING/PYTHON CODE/leadsensor.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import butter, filtfilt, medfilt
# 1. Simulate Lead (Pb²⁺) sensor data (ppm concentration)
np.random.seed(42)
time = np.linspace(0, 60, 300)  # 60 seconds, 300 samples
true_lead_ppm = 0.05 + 0.02 * np.sin(0.2 * np.pi * time / 10)  # base clean signal
noise = np.random.normal(0, 0.01, size=time.shape)  # random noise
unfiltered_data = true_lead_ppm + noise  # simulated raw sensor readings
# 2. Moving Average Filter
def moving_average(data, window_size=5):
    return np.convolve(data, np.ones(window_size) / window_size, mode='same')
moving_avg_data = moving_average(unfiltered_data, window_size=5)
# 3. Median Filter
median_filtered_data = medfilt(unfiltered_data, kernel_size=5)
# 4. Butterworth Low-Pass Filter
def butter_lowpass_filter(data, cutoff, fs, order=5):
    nyq = 0.5 * fs
    normal_cutoff = cutoff / nyq
    b, a = butter(order, normal_cutoff, btype='low', analog=False)
    return filtfilt(b, a, data)
fs = 300 / 60  # sampling frequency = 300 samples / 60 sec = 5 Hz
cutoff = 0.5   # cutoff frequency (Hz)
butter_filtered_data = butter_lowpass_filter(unfiltered_data, cutoff, fs, order=4)
# 5. Visualization
plt.figure(figsize=(12, 8))
plt.plot(time, unfiltered_data, label='Unfiltered (Raw)', color='gray', alpha=0.6)
plt.plot(time, moving_avg_data, label='Moving Average Filtered', linewidth=2)
plt.plot(time, median_filtered_data, label='Median Filtered', linewidth=2)
plt.plot(time, butter_filtered_data, label='Butterworth Filtered', linewidth=2)
plt.title("Lead (Pb²⁺) Sensor Data — Water Treatment Filtration Comparison", fontsize=14)
plt.xlabel("Time (seconds)", fontsize=12)
plt.ylabel("Lead Concentration (ppm)", fontsize=12)
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()
# 6. Comparison Summary
print("=== Lead Sensor Data Filtration Comparison Summary ===")
print(f"Unfiltered Mean Pb²⁺: {np.mean(unfiltered_data):.4f} ppm")
print(f"Moving Average Filtered Mean Pb²⁺: {np.mean(moving_avg_data):.4f} ppm")
print(f"Median Filtered Mean Pb²⁺: {np.mean(median_filtered_data):.4f} ppm")
print(f"Butterworth Filtered Mean Pb²⁺: {np.mean(butter_filtered_data):.4f} ppm")